File: cplusplus/cplusplus/AddTwoNumbers.cpp
#include <vector>
#include <unordered_map>
#include <iostream>
#include <string>
#include "ListNode.h"
using namespace std;
class AddTwoNumbers {

public:

	ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
		ListNode *head = new ListNode(0);
		if (l1 == NULL && l2 == NULL) {
			return head;
		}
		int carry = 0;
		ListNode *pointer = head;
		while (l1 != NULL || l2 != NULL) {
			int sum = 0;
			if (l1 != NULL) {
				sum += l1->val;
				l1 = l1->next;
			}
			if (l2 != NULL) {
				sum += l2->val;
				l2 = l2->next;
			}
			sum += carry;
			if (sum > 9) {
				carry = 1;
				sum -= 10;
			}
			else {
				carry = 0;
			}
			ListNode *aNode = new ListNode(sum);
			pointer->next = aNode;
			pointer = pointer->next;
		}
		if (carry != 0) {
			pointer->next = new ListNode(1);
		}
		return head->next;
	}
};

int main() {

	AddTwoNumbers s;
	ListNode *n1 = new ListNode(5);
	ListNode *n2 = new ListNode(5);
	ListNode *ret = s.addTwoNumbers(n1, n2);
	while (ret != NULL) {
		cout << ret->val << " ";
		ret = ret->next;
	}
	system("pause");
}

File: java/wenxin/leetcode/java/AddTwoNumbers_.java
package wenxin.leetcode.java;
/*
 * problem 2.
 * You are given two linked lists representing two non-negative numbers. The digits are stored in reverse order and each of their nodes contain a single digit. Add the two numbers and return it as a linked list.

Input: (2 -> 4 -> 3) + (5 -> 6 -> 4)
Output: 7 -> 0 -> 8
 */
public class AddTwoNumbers_ {
	public class ListNode {
		int val;
		ListNode next;

		ListNode(int x) {
			val = x;
			next = null;
		}
	}

	public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
		int carry = 0;
		ListNode curNode = new ListNode(0);
		ListNode first = curNode;
		while (l1 != null || l2 != null || carry != 0) {
			int sum = (l1 == null ? 0 : l1.val) + (l2 == null ? 0 : l2.val) + carry;
			ListNode n = new ListNode(sum % 10);
			curNode.next = n;
			curNode = n;
			carry = sum / 10;
			if (l1 != null)
				l1 = l1.next;
			if (l2 != null)
				l2 = l2.next;
		}
		return first.next;
	}
}
